data-prepper-plugins/opensearch/src/main/java/org/opensearch/dataprepper/plugins/sink/opensearch/OpenSearchSink.java

```diff
@@ -241,11 +250,70 @@ private void doInitializeInternal() throws IOException {
     LOG.info("Initialized OpenSearch sink");
   }
 
+  double getInvalidActionErrorsCount() {
+    return invalidActionErrorsCounter.count();
+  }
+
   @Override
   public boolean isReady() {
     return initialized;
   }
 
+  private BulkOperation getBulkOperationForAction(final String action, final SerializedJson document, final String indexName, final JsonNode jsonNode) {
+    BulkOperation bulkOperation;
+    final Optional<String> docId = document.getDocumentId();
+    final Optional<String> routing = document.getRoutingField();
+
+    if (StringUtils.equals(action, BulkAction.CREATE.toString())) {
+       final CreateOperation.Builder<Object> createOperationBuilder =
+         new CreateOperation.Builder<>()
+             .index(indexName)
+             .document(document);
+       docId.ifPresent(createOperationBuilder::id);
+       routing.ifPresent(createOperationBuilder::routing);
+       bulkOperation = new BulkOperation.Builder()
+                           .create(createOperationBuilder.build())
+                           .build();
+       return bulkOperation;
+    }
+    if (StringUtils.equals(action, BulkAction.UPDATE.toString()) ||
+        StringUtils.equals(action, BulkAction.UPSERT.toString())) {
+          final UpdateOperation.Builder<Object> updateOperationBuilder = (action.toLowerCase() == BulkAction.UPSERT.toString()) ?
+              new UpdateOperation.Builder<>()
+                  .index(indexName)
+                  .document(jsonNode)
+                  .upsert(jsonNode) :
+              new UpdateOperation.Builder<>()
+                  .index(indexName)
+                  .document(jsonNode);
+          docId.ifPresent(updateOperationBuilder::id);
+          routing.ifPresent(updateOperationBuilder::routing);
+          bulkOperation = new BulkOperation.Builder()
+                              .update(updateOperationBuilder.build())
+                              .build();
+          return bulkOperation;
```

Comment:
Just curious did you do any manual testing of these 2? Just to verify UPSERT will create or update and update will updates certain fields?

Reply:
I have added integration test cases. Please take a look and let me know.